File: dsa-problems/Stacks/solutions/javascript/STK-012-campus-expression-optimizer.js
class Solution {
  solve(expr) {
    let postfix = "";
    const ops = [];
    let redundant = 0;
    
    const prec = {
      '+': 1, '-': 1,
      '*': 2, '/': 2, '%': 2,
      '^': 3,
      '(': 0
    };
    
    // 0: Start, 1: Operand, 2: Operator, 3: Open, 4: Close
    let lastType = 0;
    
    for (let i = 0; i < expr.length; i++) {
      const c = expr[i];
      
      if (/[a-zA-Z0-9]/.test(c)) {
        if (lastType === 1 || lastType === 4) return "ERROR Invalid syntax 0";
        postfix += c;
        lastType = 1;
      } else if (c === '(') {
        if (lastType === 1 || lastType === 4) return "ERROR Invalid syntax 0";
        ops.push(c);
        lastType = 3;
      } else if (c === ')') {
        if (lastType === 0 || lastType === 2 || lastType === 3) return "ERROR Invalid syntax 0";
        
        let hasOp = false;
        while (ops.length > 0 && ops[ops.length - 1] !== '(') {
          postfix += ops.pop();
          hasOp = true;
        }
        
        if (ops.length === 0) return "ERROR Mismatched parentheses 0";
        ops.pop(); // Pop '('
        
        if (!hasOp) {
          redundant++;
        }
        lastType = 4;
      } else if (prec.hasOwnProperty(c)) {
        if (lastType === 0 || lastType === 2 || lastType === 3) return "ERROR Invalid syntax 0";
        
        while (ops.length > 0 && ops[ops.length - 1] !== '(' &&
              (prec[ops[ops.length - 1]] > prec[c] ||
              (prec[ops[ops.length - 1]] === prec[c] && c !== '^'))) {
          postfix += ops.pop();
        }
        ops.push(c);
        lastType = 2;
      } else {
        return "ERROR Invalid character 0";
      }
    }
    
    if (lastType === 0 || lastType === 2 || lastType === 3) return "ERROR Invalid syntax 0";
    
    while (ops.length > 0) {
      if (ops[ops.length - 1] === '(') return "ERROR Mismatched parentheses 0";
      postfix += ops.pop();
    }
    
    return `POSTFIX ${postfix} ${redundant}`;
  }
}

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.on("line", (line) => {
  if (line.trim() !== "") {
    const solution = new Solution();
    console.log(solution.solve(line.trim()));
    process.exit(0);
  }
});

File: dsa-problems/Stacks/solutions/java/STK-012-campus-expression-optimizer.java
import java.util.*;
import java.io.*;

class Solution {
    Map<Character, Integer> prec = new HashMap<>();
    
    public Solution() {
        prec.put('+', 1); prec.put('-', 1);
        prec.put('*', 2); prec.put('/', 2); prec.put('%', 2);
        prec.put('^', 3);
        prec.put('(', 0);
    }

    public String solve(String expr) {
        StringBuilder postfix = new StringBuilder();
        Stack<Character> ops = new Stack<>();
        int redundant = 0;
        
        // 0: Start, 1: Operand, 2: Operator, 3: Open, 4: Close
        int lastType = 0;
        
        for (int i = 0; i < expr.length(); i++) {
            char c = expr.charAt(i);
            
            if (Character.isLetterOrDigit(c)) {
                if (lastType == 1 || lastType == 4) return "ERROR Invalid syntax 0";
                postfix.append(c);
                lastType = 1;
            } else if (c == '(') {
                if (lastType == 1 || lastType == 4) return "ERROR Invalid syntax 0";
                ops.push(c);
                lastType = 3;
            } else if (c == ')') {
                if (lastType == 0 || lastType == 2 || lastType == 3) return "ERROR Invalid syntax 0";
                
                boolean hasOp = false;
                while (!ops.isEmpty() && ops.peek() != '(') {
                    postfix.append(ops.pop());
                    hasOp = true;
                }
                
                if (ops.isEmpty()) return "ERROR Mismatched parentheses 0";
                ops.pop(); // Pop '('
                
                if (!hasOp) {
                    redundant++;
                }
                lastType = 4;
            } else if (prec.containsKey(c)) {
                if (lastType == 0 || lastType == 2 || lastType == 3) return "ERROR Invalid syntax 0";
                
                while (!ops.isEmpty() && ops.peek() != '(' && 
                      (prec.get(ops.peek()) > prec.get(c) || 
                      (prec.get(ops.peek()).equals(prec.get(c)) && c != '^'))) {
                    postfix.append(ops.pop());
                }
                ops.push(c);
                lastType = 2;
            } else {
                return "ERROR Invalid character 0";
            }
        }
        
        if (lastType == 0 || lastType == 2 || lastType == 3) return "ERROR Invalid syntax 0";
        
        while (!ops.isEmpty()) {
            if (ops.peek() == '(') return "ERROR Mismatched parentheses 0";
            postfix.append(ops.pop());
        }
        
        return "POSTFIX " + postfix.toString() + " " + redundant;
    }
}

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String expr = br.readLine();
        if (expr != null) {
            Solution sol = new Solution();
            System.out.println(sol.solve(expr.trim()));
        }
    }
}
